pandas/_libs/tslibs/period.pyx

```diff
@@ -2404,6 +2405,13 @@ class Period(_Period):
                 value = str(value)
             value = value.upper()
             dt, reso = parse_time_string(value, freq)
+            try:
+                ts = Timestamp(value, freq=freq)
+                nanosecond = ts.nanosecond
+                if nanosecond != 0:
+                    reso = 'nanosecond'
+            except:
+                pass
```

Comment:
I haven't yet looked at this closely enough to give a proper review, but I'll just make a couple of comments:

- I'm almost certain that a dry except like this won't go down well with the core team. Either catch the exception, or (even better, if it's possible) don't use try-except at all;
- you write "There is for sure a better solution". Even if that's the case, it's still advisable to include a test in your PR - in fact, [it's advised](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#test-driven-development-code-writing) to write the test(s) before changing the code